packages/types-codec/src/abstract/IntBigInt.ts

```diff
@@ -0,0 +1,272 @@
+}
+
+function toPercentage (value: bigint, divisor: bigint): string {
+  return `${(Number((value * MUL_P) / divisor) / 100).toFixed(2)}%`;
+}
+
+/** @internal */
+function decodeBigInt (value: Exclude<AnyNumber, Uint8Array> | Record<string, string>, isNegative: boolean): bigint {
+  if (isBigInt(value)) {
+    return value;
+  } else if (isNumber(value)) {
+    if (!Number.isInteger(value) || value > Number.MAX_SAFE_INTEGER || value < Number.MIN_SAFE_INTEGER) {
+      throw new Error('Number needs to be an integer <= Number.MAX_SAFE_INTEGER, i.e. 2 ^ 53 - 1');
+    }
+
+    return BigInt(value);
+  } else if (isString(value)) {
+    if (isHex(value, -1, true)) {
+      return hexToBigInt(value, { isLe: false, isNegative });
+    }
+
+    if (value.includes('.') || value.includes(',') || value.includes('e')) {
+      throw new Error('String should not contain decimal points or scientific notation');
+    }
+
+    return BigInt(value);
+  } else if (isBn(value)) {
+    return BigInt(value.toString());
+  } else if (isObject(value)) {
+    if (isToBigInt(value)) {
+      return value.toBigInt();
+    }
+
+    // Allow the construction from an object with a single top-level key. This means that
+    // single key objects can be treated equivalently to numbers, assuming they meet the
+    // specific requirements. (This is useful in Weights 1.5 where Objects are compact)
+    const keys = Object.keys(value);
+
+    if (keys.length !== 1) {
+      throw new Error('Unable to construct number from multi-key object');
+    }
+
+    return decodeBigInt(value[keys[0]], isNegative);
+  }
+
+  throw new Error(`Unable to create BigInt from unknown type ${typeof value}`);
+}
+
+/**
+ * @name AbstractBigInt
+ * @ignore
+ * @noInheritDoc
+ */
+export abstract class AbstractBigInt extends AbstractObject<bigint> implements INumber {
+  readonly isUnsigned: boolean;
+
+  readonly #bitLength: UIntBitLength;
+  readonly #bitLengthInitial: number;
+  readonly #encodedLength: number;
+
+  constructor (registry: Registry, value: AnyNumber = 0, bitLength: UIntBitLength = DEFAULT_UINT_BITS, isSigned = false) {
+    super(
+      registry,
+      // shortcut isU8a as used in SCALE decoding
+      isU8a(value)
+        ? u8aToBigInt(value.subarray(0, bitLength / 8), { isLe: true, isNegative: isSigned })
+        : decodeBigInt(value, isSigned),
+      bitLength / 8
+    );
+
+    this.#bitLength = bitLength;
+    this.#bitLengthInitial = this.$.toString(2).length;
+    this.#encodedLength = this.#bitLength / 8;
+    this.isUnsigned = !isSigned;
+
+    const isNegative = this.$ < _0n;
+    const maxBits = bitLength - (isSigned && !isNegative ? 1 : 0);
+
+    if (isNegative && !isSigned) {
+      throw new Error(`${this.toRawType()}: Negative number passed to unsigned type`);
+    } else if (this.#bitLengthInitial > maxBits) {
+      throw new Error(`${this.toRawType()}: Input too large. Found input with ${this.#bitLengthInitial} bits, expected ${maxBits}`);
+    }
```

Comment:
It would be a bit more efficient to place validation before creating instance of a class because in case invalid state no extra operations for fill instance properties.

```suggestion
    if (this.$ < _0n && !isSigned) {
      throw new Error(`${this.toRawType()}: Negative number passed to unsigned type`);
    }
    this.#bitLengthInitial = this.$.toString(2).length;
    if (this.#bitLengthInitial >= maxBits) {
      throw new Error(`${this.toRawType()}: Input too large. Found input with ${this.#bitLengthInitial} bits, greater ${maxBits}`);
    }
    this.#bitLength = bitLength;
    this.#encodedLength = this.#bitLength / 8;
    this.isUnsigned = !isSigned;
```